src/Evolution/Actions/BackgroundGrVars.hpp

```diff
@@ -0,0 +1,184 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <optional>
+#include <utility>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/Variables.hpp"
+#include "DataStructures/VariablesTag.hpp"
+#include "Domain/Block.hpp"
+#include "Domain/Creators/Tags/Domain.hpp"
+#include "Domain/Domain.hpp"
+#include "Domain/Structure/BlockId.hpp"
+#include "Domain/Structure/Element.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Domain/Tags.hpp"
+#include "Evolution/Initialization/InitialData.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Parallel/AlgorithmExecution.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MutateAssign.hpp"
+#include "PointwiseFunctions/AnalyticData/Tags.hpp"
+#include "PointwiseFunctions/InitialDataUtilities/InitialData.hpp"
+#include "PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/CallWithDynamicType.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace evolution::Actions {
+
+/*!
+ * \brief Allocate or assign background general relativity quantities needed
+ * for evolution systems that run on a fixed spacetime e.g. ValenciaDivclean,
+ * ForceFree.
+ *
+ * If a template parameter `used_in_evolution_phase` == true, this action
+ * mutates (updates) the value of background metric variables in box.
+ *
+ * Uses:
+ *  - DataBox:
+ *    * `evolution::initial_data::Tags::InitialData` or
+ *       `::Tags::AnalyticSolutionOrData`
+ *    * `domain::Tags::Coordinates<Dim, Frame::Inertial>`
+ *    * `domain::Tags::Element<Dim>`
+ *    * `domain::Tags::Domain<Dim>`
+ *    * `::Tags::Time`
+ *
+ * DataBox changes:
+ *  - Adds: nothing
+ *  - Removes : nothing
+ *  - Modifies :
+ *    * system::spacetime_variables_tag
+ *    * system::flux_spacetime_variables_tag
+ *    * system::inverse_spatial_metric_tag
+ *
+ * If a template parameter `used_in_evolution_phase` == false, this action can
+ * be used for initialization. Allocate and assign the values of background
+ * metric variables into DataBox.
+ *
+ * Uses:
+ *  - DataBox:
+ *    * `evolution::initial_data::Tags::InitialData` or
+ *       `::Tags::AnalyticSolutionOrData`
+ *    * `domain::Tags::Coordinates<Dim, Frame::Inertial>`
+ *    * `domain::Tags::Mesh<Dim>`
+ *    * `::Tags::Time`
+ *
+ * DataBox changes:
+ *  - Adds:
+ *    * system::spacetime_variables_tag
+ *    * system::flux_spacetime_variables_tag
+ *    * system::inverse_spatial_metric_tag
+ *  - Removes : nothing
+ *  - Modifies : nothing
+ *
+ */
+template <typename System, bool used_in_evolution_phase>
+struct BackgroundGrVars {
```

Comment:
A few questions for our meeting later:

1. I think you can get rid of the `used_in_evolution_phase` and just check if the size of the variables are non-zero.
2. This seems to only be updating the active tags, but always using the GR coordinates. Naively I expect this to not work with subcell. I _think_ what works is if you forward declare the subcell tags and then do the `if constexpr (tag_is_retrievable) {` that'll work with just a forward decl. If not, we (I) need to figure out how to make that work, because it should. Then you can have an internal check basically for "Could I be doing subcell? If so, then check what the active grid is, get the right coordinates, etc. else just do the DG case"

So as-is, I expect the code to actually not work correctly because 1) the face tags aren't updated and 2) the cell-centered tags are always set to the DG grid. I think this code here is different from what you were sharing in our call, though.